android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/Utils.kt

```diff
@@ -11,80 +11,59 @@ private const val SEARCH_TYPE_SAP_FOLLOW_ON = "sap-follow-on"
-        ) {
-            code = uri.toString().substringAfter("from=", "")
-                .substringBefore("/", "")
-        }
-
-        // For Bug 1751920
-        if (!validCodeSet.contains(code)) {
-            code = "other"
-        }
-
-        // Glean doesn't allow code starting with a figure
-        if (code != null && code.isNotEmpty()) {
-            val codeStart = code.first()
-            if (codeStart.isDigit()) {
-                code = "_$code"
+    var code: String? = "none"
+
+    if (provider.codeParamName.isNotEmpty()) {
+        code = uri.getQueryParameter(provider.codeParamName)
+        if (code != null) {
+            // The code is only included if it matches one of the specific ones.
+            if (provider.taggedCodes.contains(code)) {
+                type = SEARCH_TYPE_SAP
+                if (provider.followOnParamNames?.any { p -> paramSet.contains(p) } == true) {
+                    type = SEARCH_TYPE_SAP_FOLLOW_ON
+                }
+            } else if (provider.organicCodes?.contains(code) == true) {
+                type = SEARCH_TYPE_ORGANIC
+            } else if (provider.expectedOrganicCodes?.contains(code) == true) {
+                code = "none"
+            } else {
+                code = if (code.first().isDigit()) "_$code" else "other"
             }
-        }
-
-        // Try cookies first because Bing has followOnCookies and valid code, but no
-        // followOnParams => would tracks organic instead of sap-follow-on
-        if (provider.followOnCookies.isNotEmpty()) {
-            // Checks if engine contains a valid follow-on cookie, otherwise return default
+        } else if (provider.followOnCookies != null) {
+            // Try cookies first because Bing has followOnCookies and valid code, but no
+            // followOnParams => would track organic instead of sap-follow-on
             getTrackKeyFromCookies(provider, uri, cookies)?.let {
                 return it.createTrackKey()
             }
         }
 
         // For Bing if it didn't have a valid cookie and for all the other search engines
-        if (hasValidCode(uri.getQueryParameter(provider.codeParam), provider)) {
+        if (hasValidCode(uri.getQueryParameter(provider.codeParamName), provider)) {
             var channel = uri.getQueryParameter(CHANNEL_KEY)
 
             // For Bug 1751955
             if (!validChannelSet.contains(channel)) {
                 channel = null
             }
-
-            val type = getSapType(provider.followOnParams, paramSet)
-            return TrackKeyInfo(provider.name, type, code, channel).createTrackKey()
+            return TrackKeyInfo(provider.telemetryId, type, code, channel).createTrackKey()
         }
     }
-
-    // Default to organic search type if no code parameter was found.
-    return TrackKeyInfo(provider.name, SEARCH_TYPE_ORGANIC, code).createTrackKey()
+    return TrackKeyInfo(provider.telemetryId, type, code).createTrackKey()
```

Comment:
I think it could potentially help short-circuit some of the understanding of this function to move this to the top, for example:

```
if (provider.codeParamName.isEmpty()) {
    return TrackKeyInfo(provider.telemetryId, SEARCH_TYPE_ORGANIC, "none").createTrackKey()
} else {
    var type = SEARCH_TYPE_ORGANIC
    var code = uri.getQueryParameter(provider.codeParamName)
    ...
}

Reply:
This was causing for us to have 4 return statements and apparently our detekt doesn't allow that